examples/structured_data/movielens_recommendations_transformers.py

```diff
@@ -0,0 +1,606 @@
+    encoded_sequence_movies_with_poistion_and_rating = layers.Multiply()(
+        [(encoded_sequence_movies + encodded_positions), sequence_ratings]
+    )
+    #################################################################
+
+    # Construt the transformer inputs.
+    for encoded_movie in tf.unstack(
+        encoded_sequence_movies_with_poistion_and_rating, axis=1
+    ):
+        encoded_transformer_features.append(tf.expand_dims(encoded_movie, 1))
+    encoded_transformer_features.append(encoded_target_movie)
+
+    encoded_transformer_features = layers.concatenate(
+        encoded_transformer_features, axis=1
+    )
+
+    return encoded_transformer_features, encoded_other_features
+
+
+"""
+## Create a BST model
+"""
+
+include_user_id = False
+include_user_features = False
+include_movie_features = False
+
+hidden_units = [256, 128]
+dropout_rate = 0.1
+num_heads = 3
+
+
+def create_model():
+    inputs = create_model_inputs()
+    transformer_features, other_features = encode_input_features(
+        inputs, include_user_id, include_user_features, include_movie_features
+    )
+
+    # Create a multi-headed attention layer.
+    attention_output = layers.MultiHeadAttention(
+        num_heads=num_heads, key_dim=transformer_features.shape[2], dropout=dropout_rate
+    )(transformer_features, transformer_features)
+
+    # Transformer block.
+    attention_output = layers.Dropout(dropout_rate)(attention_output)
+    x1 = layers.Add()([transformer_features, attention_output])
+    x1 = layers.LayerNormalization()(x1)
+    x2 = layers.LeakyReLU()(x1)
+    x2 = layers.Dense(units=x2.shape[-1])(x2)
+    x2 = layers.Dropout(dropout_rate)(x2)
+    transformer_features = layers.Add()([x1, x2])
+    transformer_features = layers.LayerNormalization()(transformer_features)
+    features = layers.Flatten()(transformer_features)
+
+    # Included the other features.
+    if other_features is not None:
+        features = layers.concatenate(
+            [features, layers.Reshape([other_features.shape[-1]])(other_features)]
+        )
+
+    # Fully-connected layers.
+    for num_units in hidden_units:
+        features = layers.Dense(num_units)(features)
+        features = layers.BatchNormalization()(features)
+        features = layers.LeakyReLU()(features)
+        features = layers.Dropout(dropout_rate)(features)
+
+    outputs = layers.Dense(units=1)(features)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    return model
+
+
+model = create_model()
+keras.utils.plot_model(model, show_shapes=True)
+
+"""
+## Run training and evaluation experiment
+"""
+
+learning_rate = 0.01
+batch_size = 265
+num_epochs = 5
+
+
+def run_experiment(model):
+
+    model.compile(
+        optimizer=keras.optimizers.Adagrad(learning_rate=learning_rate),
+        loss=keras.losses.MeanSquaredError(),
+        metrics=[keras.metrics.MeanAbsoluteError()],
+    )
+
+    train_dataset = get_dataset_from_csv(
```

Comment:
Suggestion: add a comment

```suggestion
    # Load the training set from the DataFrame.
    train_dataset = get_dataset_from_csv(
```